### PR TITLE
Add cv_results_ to OptunaSearchCV

### PR DESCRIPTION
**Pull request by jckkvs**
## Motivation
This is to enhance compatibility with [sklearn.GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

The user_attrs is already implemented in Optuna, to obtain the score of each trial, but cv_results_enable to obtain the scores of all trials at once.

In addition,
we can easily calculate the bias_correction proposed by Tibshirani and Tibshirani by cv_results_　(https://tibshirani.su.domains/ftp/AOAS224.pdf).

## Description of the changes

Add cv_results_ to OptunaSearchCV

The variable names  ```cv_results_dict_in_list``` , ```cv_results_list_in_dict``` are very long, so we can change them according to optuna's conventions.

### CODE REVIEW

File: optuna/study/study.py [hunk 1]
```diff
@@ -174,6 +174,20 @@ def best_trials(self) -> List[FrozenTrial]:
...
+
+        Returns:
+            A list of results of all trials.
+        """
+        cv_results_dict_in_list = [trial_.user_attrs for trial_ in self.trials]
+        cv_results_list_in_dict = {
+            key: [dict_[key] for dict_ in cv_results_dict_in_list]
+            for key in cv_results_dict_in_list[0]
+        }
+        return cv_results_list_in_dict
```
Comment (toshihikoyanase) [inline optuna/study/study.py:189]: Could you consider moving this implementation to `optuna/integration/sklearn.py`?
`Trial.user_attrs` can hold a variety of data, not just CV results, so the current implementation might be specific to `OptunaSearchCV`.

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your PR. I'm a bit hesitant about introducing a new method to the `Study` class, as this functionality seems to be specific to `OptunaSearchCV`. Please check the following comment.

**Comment by toshihikoyanase** [comment]
@Alnusjaponica Could you join the review, please?

**Review by Alnusjaponica — CHANGES_REQUESTED** [review]
@jckkvs You need to test the behavior of `OptunaSearchCV.cv_results_` too.

File: optuna/integration/sklearn.py [hunk 2]
```diff
@@ -563,6 +563,19 @@ def classes_(self) -> OneDimArrayLikeType:
 
         return self.best_estimator_.classes_
 
+    @property
+    def cv_results_(self) -> Dict[str, Any]:
+        """Return the dictionary has list of all trials.
+        Returns:
+            A list of results of all trials.
+        """
```
Comment (toshihikoyanase) [inline optuna/integration/sklearn.py:571]: Sorry for the delayed response. Could you update the docstrings? I don't think this attribute returns all trials (i.e., `list[FrozenTrial]`). In my understanding, it returns a list of CV results corresponding to all the trials.

File: optuna/integration/sklearn.py [hunk 3]
```diff
@@ -563,6 +563,19 @@ def classes_(self) -> OneDimArrayLikeType:
...
+        cv_results_dict_in_list = [trial_.user_attrs for trial_ in self.trials_]
+        cv_results_list_in_dict = {
+            key: [dict_[key] for dict_ in cv_results_dict_in_list]
+            for key in cv_results_dict_in_list[0]
```
Comment (Alnusjaponica) [inline optuna/integration/sklearn.py:575]: Is this safe when `len(cv_results_dict_in_list)==0`?

File: optuna/integration/sklearn.py [hunk 4]
```diff
@@ -563,6 +563,22 @@ def classes_(self) -> OneDimArrayLikeType:
 
         return self.best_estimator_.classes_
 
+    @property
+    def cv_results_(self) -> Dict[str, Any]:
+        """Return the dictionary has list of CV results of all trials.
+        Returns:
+            A list of CV results of all trials.
+        """
```
Comment (toshihikoyanase) [inline optuna/integration/sklearn.py:571]: These lines seem to be duplicated. How about unifying it like the other properties?

```suggestion
        """A dictionary mapping a metric name to a list of Cross-Validation results of all trials."""
```

File: optuna/integration/sklearn.py (hunk 2 above)
Comment (toshihikoyanase) [reply optuna/integration/sklearn.py]: Thanks!

File: optuna/study/study.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/study/study.py]: > your proposal contains typo

Oops! Sorry for bothering you. Please ignore my comment and overwrite it with your change.

**Review by toshihikoyanase — APPROVED** [review]
I'm sorry for the delayed response. LGTM!
@Alnusjaponica Please merge after your review.

**Review by Alnusjaponica — APPROVED** [review]
Apologies for the delay. LGTM too.
@toshihikoyanase Could you merge this PR?

**Comment by toshihikoyanase** [comment]
@Alnusjaponica Thank you for your review. Let me merge this PR.

**Pull request merged**